### PR TITLE
IrrCompileConfig change

### PR DESCRIPTION
**Pull request by devshgraphicsprogramming**
@Crisspl let me know when this is ready

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
mention #155

**Comment by devshgraphicsprogramming** [comment]
@Crisspl nice, but we need to review that everything that needs to be configurable/installable, is so.

My worry is that some of the `-D` defines might trip up projects using irrBaW
https://github.com/buildaworldnet/IrrlichtBAW/blob/facc5d9225baba6d0056f6a60be7912e8a7ade53/CMakeLists.txt#L96

**Comment by devshgraphicsprogramming** [comment]
@Crisspl you need to merge my master PR into your master as well as cherrypick a few commits from my sRGB branch (@manhnt9's Linux fixes)

File: include/BuildConfigOptions.h.in [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#ifndef __IRR_BUILD_CONFIG_OPTIONS_H_INCLUDED__
```
Comment (devshgraphicsprogramming) [inline include/BuildConfigOptions.h.in]: @Crisspl lovely.

**Comment by devshgraphicsprogramming** [comment]
@Crisspl anything else we need to do?

**Comment by Crisspl** [comment]
>anything else we need to do?

if i remember correctly, you wanted to have `_IRR_SERVER` and `_IRR_DEBUG` as "installables". And i don't really know how to do this especially in multi-config IDEs.

**Comment by devshgraphicsprogramming** [comment]
does `_IRR_SERVER` influence anything on the outside?

File: CMakeLists.txt [hunk 2]
```diff
@@ -30,8 +30,9 @@ macro(irr_adjust_flags)
 		add_compile_options(
 			-Wextra
 			-Wno-unused-parameter
+			-Wno-unused-but-set-parameter
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:33]: rationale?

File: include/SIMDswizzle.h [hunk 3]
```diff
@@ -274,6 +274,12 @@ class IRR_FORCE_EBO SIMD_32bitSwizzleAble
 
 #define FAST_FLOAT_SHUFFLE(X,Y) _mm_castsi128_ps(_mm_shuffle_epi32(_mm_castps_si128(X),Y))
 
+#ifdef __GNUC__
+// warning: ignoring attributes on template argument �__m128i {aka __vector(2) long long int}� [-Wignored-attributes] (etc...)
```
Comment (devshgraphicsprogramming) [inline include/SIMDswizzle.h:278]: I thought we already had "Wnoerrr=ignored-attributes"

**Comment by devshgraphicsprogramming** [comment]
@Crisspl did you add the `$(OutDir)` MSVC debug working directory CMAKE option to the make_example cmake macro/function?
https://cmake.org/cmake/help/v3.8/prop_tgt/VS_DEBUGGER_WORKING_DIRECTORY.html

**Comment by devshgraphicsprogramming** [comment]
> if i remember correctly, you wanted to have _IRR_SERVER and _IRR_DEBUG as "installables". And i don't really know how to do this especially in multi-config IDEs.

I think we covered `_IRR_SERVER` in our private chat.

for different Configurations (Release vs. Debug) we obviously install to different paths/with different names.

So to refresh, recap and adjust:

Out default `CMAKE_INSTALL_PREFIX = ./install/$Platform` (double check this)

**NEW: Always add `/debug` to CMAKE_INSTALL_PREFIX if the build type is Debug**

With usual default of `/usr/local` this would evaluate to
`./install/$Platform`
and
`./install/$Platform/debug`
respectively.

Then we install as follows:
1. Headers go into `$CMAKE_INSTALL_PREFIX/include`
1. Built Libraries go into `$CMAKE_INSTALL_PREFIX/lib`

For system-wide install, this expands to 

Debug Lib -> /usr/local/debug/lib/
Debug Headers -> /usr/local/debug/include/

Release Lib -> /usr/local/lib/
Release Headers -> /usr/local/include/

I base this off the standard linux install and packaging conventions.

File: include/SIMDswizzle.h (hunk 3 above)
Comment (devshgraphicsprogramming) [reply include/SIMDswizzle.h:278]: `-Wno-error=ignored-attributes` doesn't take care of this?

**Comment by devshgraphicsprogramming** [comment]
Is the following done?
https://github.com/buildaworldnet/IrrlichtBAW/issues/155#issuecomment-483610248

**Comment by devshgraphicsprogramming** [comment]
> > if i remember correctly, you wanted to have _IRR_SERVER and _IRR_DEBUG as "installables". And i don't really know how to do this especially in multi-config IDEs.
> 
> I think we covered `_IRR_SERVER` in our private chat.
> 
> for different Configurations (Release vs. Debug) we obviously install to different paths/with different names.
> 
> So to refresh, recap and adjust:
> 
> Out default `CMAKE_INSTALL_PREFIX = ./install/$Platform` (double check this)
> 
> **NEW: Always add `/debug` to CMAKE_INSTALL_PREFIX if the build type is Debug**
> 
> With usual default of `/usr/local` this would evaluate to
> `./install/$Platform`
> and
> `./install/$Platform/debug`
> respectively.
> 
> Then we install as follows:
> 
> 1. Headers go into `$CMAKE_INSTALL_PREFIX/include`
> 2. Built Libraries go into `$CMAKE_INSTALL_PREFIX/lib`
> 
> For system-wide install, this expands to
> 
> Debug Lib -> /usr/local/debug/lib/
> Debug Headers -> /usr/local/debug/include/
> 
> Release Lib -> /usr/local/lib/
> Release Headers -> /usr/local/include/
> 
> I base this off the standard linux install and packaging conventions.

The final thing to think about is how this all will fit in with #91 ?

I personally think nothing changes for shared library builds except for macros for `IRRCALLCONV` and `IRRLICHT_API` in front of function definitions and declaring `IRRLICHT_EXPORTS` in build params.

Installing `_IRR_STATIC_LIB` may seem like a nice idea but the app linking the library needs to link to .a or .so (.lib or .dll) so it could just define `_IRR_STATIC_LIB` when appropriate anyway.

This only creates a new output binary and keeps headers identical.

**Comment by Crisspl** [comment]
>Is the following done?

yes

Ok, so `_IRR_DEBUG` goes to config-header file as "installable", right? I think I finaly have an idea how to do this to be usable for devs and users

**Comment by devshgraphicsprogramming** [comment]
yes its installable.

File: include/SIMDswizzle.h (hunk 3 above)
Comment (Crisspl) [reply include/SIMDswizzle.h:278]: hm apparently not

File: cmake/common.cmake [hunk 4]
```diff
@@ -44,12 +44,29 @@ macro(irr_create_executable_project _EXTRA_SOURCES _EXTRA_OPTIONS)
 	endif()
 endmacro()
 
+
+function(irr_get_conf_dir _OUTVAR _CONFIG)
+	string(TOLOWER ${_CONFIG} CONFIG)
+	set(${_OUTVAR} "${CMAKE_SOURCE_DIR}/include/irr/config/${CONFIG}" PARENT_SCOPE)
```
Comment (devshgraphicsprogramming) [inline cmake/common.cmake:51]: why special directory for config?

**Comment by devshgraphicsprogramming** [comment]
> 
> 
> @Crisspl did you add the `$(OutDir)` MSVC debug working directory CMAKE option to the make_example cmake macro/function?
> https://cmake.org/cmake/help/v3.8/prop_tgt/VS_DEBUGGER_WORKING_DIRECTORY.html

friendly reminder

File: cmake/common.cmake [hunk 5]
```diff
@@ -40,6 +40,7 @@ macro(irr_create_executable_project _EXTRA_SOURCES _EXTRA_OPTIONS)
 			PROPERTIES
 			RUNTIME_OUTPUT_DIRECTORY_DEBUG "${PROJECT_SOURCE_DIR}/bin"
 			RUNTIME_OUTPUT_DIRECTORY_RELEASE "${PROJECT_SOURCE_DIR}/bin"
+			VS_DEBUGGER_WORKING_DIRECTORY "${PROJECT_SOURCE_DIR}/bin" # seems like has no effect
```
Comment (devshgraphicsprogramming) [inline cmake/common.cmake:43]: if you dlete cmake cache and reconfigure+regenerate?

File: include/SCompoundCollider.h [hunk 6]
```diff
@@ -71,6 +71,8 @@ class SCompoundCollider : public IReferenceCounted
                             tmp->drop();
                         }
                         break;
+                    case SCollisionShapeDef::ECST_COUNT:
+                        break;
```
Comment (devshgraphicsprogramming) [inline include/SCompoundCollider.h:76]: that shouldn't happen so there should be an `assert(false)` before the break

File: source/Irrlicht/COpenGLCubemapTexture.h [hunk 7]
```diff
@@ -57,6 +57,8 @@ class COpenGLCubemapTexture : public COpenGLFilterableTexture
             case ECMF_POSITIVE_Z:
                 return GL_TEXTURE_CUBE_MAP_POSITIVE_Z;
                 break;
+            case ECMF_COUNT:
```
Comment (devshgraphicsprogramming) [inline source/Irrlicht/COpenGLCubemapTexture.h:60]: again assert

File: cmake/common.cmake (hunk 5 above)
Comment (Crisspl) [reply cmake/common.cmake:43]: tried, nothing anyway

**Pull request merged**